CRM/Upgrade/Incremental/php/FiveTwelve.php

```diff
@@ -84,4 +96,72 @@ public function setPostUpgradeMessage(&$postUpgradeMessage, $rev) {
   //   return TRUE;
   // }
 
+  /**
+   * Upgrade function.
+   *
+   * @param string $rev
+   */
+  public function upgrade_5_12_alpha1($rev) {
+    $this->addTask(ts('Upgrade DB to %1: SQL', array(1 => $rev)), 'runSql', $rev);
+    $this->addTask('Migrate to utf8mb4', 'migrateUtf8mb4');
+  }
+
+  /**
+   * Performs the utf8mb4 migration.
+   */
+  public static function migrateUtf8mb4(CRM_Queue_TaskContext $ctx) {
+    $dao = new CRM_Core_DAO();
+    $database = $dao->_database;
+    CRM_Core_DAO::executeQuery("ALTER DATABASE $database CHARACTER SET = utf8mb4 COLLATE = utf8mb4_unicode_ci");
+    $dao = CRM_Core_DAO::executeQuery("SHOW TABLE STATUS WHERE (Engine = 'InnoDB' OR Engine = 'ARCHIVE') AND Name LIKE 'civicrm\_%'");
+    while ($dao->fetch()) {
+      $tables[$dao->Name] = $dao->Engine;
+    }
+    $dsn = defined('CIVICRM_LOGGING_DSN') ? DB::parseDSN(CIVICRM_LOGGING_DSN) : DB::parseDSN(CIVICRM_DSN);
+    $logging_database = $dsn['database'];
+    $dao = CRM_Core_DAO::executeQuery("SHOW TABLE STATUS FROM `$logging_database` WHERE (Engine = 'InnoDB' OR Engine = 'ARCHIVE') AND Name LIKE 'log\_civicrm\_%'");
```

Comment:
We started using TokuDB for log tables recently (via `alterLogTables` hook), this query wouldn't match those. Does `... WHERE Comment != 'VIEW' AND Name LIKE 'log\_civicrm\_%'` seem reasonable?

Reply:
I just don't want to include any MyISAM tables that could be present, so maybe we can say `WHERE Engine <> 'MyISAM'`